File: Sample Code/ObjcSampleCode/DJISdkDemo/Demo/Bluetooth/BluetoothConnectorViewController.m
//
//  BluetoothConnectorViewController.m
//  DJISdkDemo
//
//  Copyright © 2016 DJI. All rights reserved.
//

#import "BluetoothConnectorViewController.h"
#import <Corebluetooth/CoreBluetooth.h>
#import "DemoUtility.h"
#import <DJISDK/DJISDK.h>

static NSString* const BluetoothCellReuseKey = @"BluetoothCellReuseKey";

@interface BluetoothConnectorViewController ()<UITableViewDelegate, UITableViewDataSource, DJIBluetoothProductConnectorDelegate>

@property (weak, nonatomic) IBOutlet UIButton *searchButton;
@property (weak, nonatomic) IBOutlet UIButton *connectionButton;
@property (weak, nonatomic) IBOutlet UITableView *tableView;

@property(nonatomic, strong) NSMutableArray<CBPeripheral *> *bluetoothProducts;
@property(nonatomic, assign) NSUInteger selectedIndex;
@property(nonatomic, weak, readonly) DJIBluetoothProductConnector* bluetoothConnector;

@end

@implementation BluetoothConnectorViewController

-(DJIBluetoothProductConnector *)bluetoothConnector {
    return [DJISDKManager bluetoothConnector];
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    _tableView.delegate = self;
    _tableView.dataSource = self;

    self.bluetoothProducts = [NSMutableArray new];
    self.selectedIndex = 0;
    self.bluetoothConnector.delegate = self;
    [self updateConnectionButtonUI];
}

- (IBAction)onSearchBluetoothButtonClicked:(id)sender
{
    [self.bluetoothConnector searchBluetoothProductsWithCompletion:^(NSError * _Nullable error) {
        if (error)
            ShowResult(@"Search Bluetooth product failed:%@", error.description);
    }];
}

-(void)updateConnectionButtonUI {
    if ([self isBluetoothProductConnected]) {
        [self.connectionButton setTitle:@"Disconnect" forState:UIControlStateNormal];
    }
    else {
        [self.connectionButton setTitle:@"Connect" forState:UIControlStateNormal];
    }
}

- (IBAction)onConnectionButtonClicked:(id)sender
{
    if ([self isBluetoothProductConnected]) {
        [self disconnectBluetooth];
    }
    else {
        [self connectBluetooth];
    }
}

-(void)connectBluetooth {
    WeakRef(target);
    if (!self.bluetoothProducts || self.bluetoothProducts.count == 0) {
        ShowResult(@"No Bluetooth products found. ");
        return;
    }

    [self.bluetoothConnector connectProduct:[self.bluetoothProducts objectAtIndex:self.selectedIndex] withCompletion:^(NSError * _Nullable error) {
        if (error) {
            NSLog(@"connect error:%@", error.description);
        } else {
            [target.bluetoothProducts removeAllObjects];
            [target.tableView reloadData];
            [target.connectionButton setTitle:@"Disconnect" forState:UIControlStateNormal];
        }
    }];
}

- (void)disconnectBluetooth
{
    WeakRef(target);
    [self.bluetoothConnector disconnectProductWithCompletion:^(NSError * _Nullable error) {
        if (error) {
            NSLog(@"disconnect error:%@", error.description);
        } else {
            [target.connectionButton setTitle:@"Connect" forState:UIControlStateNormal];
        }
    }];
}

-(BOOL)isBluetoothProductConnected {
    DJIBaseProduct *product = [DJISDKManager product];
    if (product) {
        if ([product.model isEqual:DJIHandheldModelNameOsmoMobile]) {
            return YES;
        }
    }
    return NO;
}

#pragma mark DJIBluetoothProductConnectorDelegate

- (void)connectorDidFindProducts:(NSArray<CBPeripheral *> *)peripherals
{
    self.bluetoothProducts = [NSMutableArray arrayWithArray:peripherals];
    [self.tableView reloadData];
}


#pragma mark UITableViewDelegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    self.selectedIndex = indexPath.row;
}

#pragma mark UITableViewDataSource

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self.bluetoothProducts.count;
}

- (UITableViewCell*)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell* cell = [tableView dequeueReusableCellWithIdentifier:BluetoothCellReuseKey];
    if (cell == nil) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:BluetoothCellReuseKey];
    }
    
    CBPeripheral* peripheral = [self.bluetoothProducts objectAtIndex:indexPath.row];
    if (peripheral) {
        cell.textLabel.text = peripheral.name ? peripheral.name : @"null";
        if (peripheral.state == CBPeripheralStateConnected) {
            cell.textLabel.backgroundColor = [UIColor colorWithRed:0 green:0 blue:1 alpha:0.2];
        } else {
            cell.textLabel.backgroundColor = [UIColor whiteColor];
        }
    }
    return cell;
}

@end

File: Sample Code/SwiftSampleCode/DJISDKSwiftDemo/Components/Bluetooth/BluetoothConnectorViewController.swift
//
//  BluetoothConnectorViewController.swift
//  DJISDKSwiftDemo
//
//  Created by DJI on 16/9/1.
//  Copyright © 2016年 DJI. All rights reserved.
//


import DJISDK
import CoreBluetooth
fileprivate func < <T : Comparable>(lhs: T?, rhs: T?) -> Bool {
  switch (lhs, rhs) {
  case let (l?, r?):
    return l < r
  case (nil, _?):
    return true
  default:
    return false
  }
}


class BluetoothConnectorViewController: DJIBaseViewController, UITableViewDelegate, DJIBluetoothProductConnectorDelegate, UITableViewDataSource {

    @IBOutlet weak var searchButton: UIButton!
    @IBOutlet weak var connectButton: UIButton!
    @IBOutlet weak var bluetoothDevicesTableView: UITableView!
    var bluetoothProducts: [CBPeripheral]?
    var selectedIndex: IndexPath!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.bluetoothProducts = []
        self.updateConnectButtonState()
        bluetoothConnector()?.delegate = self
    }
    
    func bluetoothConnector() -> DJIBluetoothProductConnector? {
        return DJISDKManager.bluetoothConnector()
    }
    
    @IBAction func onBluetoothSearchButtonClicked(_ sender: AnyObject) {
        self.bluetoothConnector()?.searchBluetoothProducts { [weak self](error:Error?) in
            if let _ = error {
                self!.showAlertResult("Search Bluetooth product failed:\(error!)")
            }
        }
    }
    
    @IBAction func onBluetoothConnectButtonClicked(_ sender: AnyObject) {
        if isBluetoothProductConnected() {
            self.disconnectBluetooth();
        } else {
            self.connectBluetooth();
        }
    }
    
    func updateConnectButtonState() -> Void {
        if isBluetoothProductConnected() {
            self.connectButton.setTitle("Disconnect", for: UIControlState())
        } else {
            self.connectButton.setTitle("Connect", for: UIControlState())
        }
    }
    
    func isBluetoothProductConnected() -> Bool {
        let product: DJIBaseProduct? = ConnectedProductManager.sharedInstance.connectedProduct
        if product != nil {
            if (product?.model == DJIHandheldModelNameOsmoMobile) {
                return true;
            }
        }
        return false;
    }
    
    func connectBluetooth() -> Void {
        
        if self.bluetoothProducts?.isEmpty == true ||
           self.selectedIndex == nil {
            return
        }
        
        let curSelectedPer = self.bluetoothProducts?[self.selectedIndex.row]
        
        self.bluetoothConnector()?.connectProduct(curSelectedPer) { [weak self](error:Error?) in
            if let _ = error {
                self!.showAlertResult("Connect Bluetooth product failed:\(error!)")
            } else {
                self!.bluetoothProducts?.removeAll();
                self!.bluetoothDevicesTableView.reloadData()
                self!.connectButton.setTitle("Disconnect", for: UIControlState())
            }
        }
    }
    
    func disconnectBluetooth() -> Void {
        self.bluetoothConnector()?.disconnectProduct { [weak self](error:Error?) in
            if let _ = error {
                self!.showAlertResult("Disconnect Bluetooth product failed:\(error!)")
            } else {
                self!.connectButton.setTitle("Connect", for: UIControlState())
                self!.bluetoothDevicesTableView.reloadData()
            }
        }
    }
    
    
    func connectorDidFindProducts(_ peripherals: [CBPeripheral]?) {
        self.bluetoothProducts = peripherals
        self.bluetoothDevicesTableView.reloadData()
    }
    
    
    //UITableView Delegate Methods
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return (self.bluetoothProducts?.count)!
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let CellIdentifier: String = "BluetoothCellReuseKey"
        var cell: UITableViewCell? = tableView.dequeueReusableCell(withIdentifier: CellIdentifier)
        if cell == nil {
            cell = UITableViewCell(style: .default, reuseIdentifier: CellIdentifier)
        }
        
        let peripheral = self.bluetoothProducts![(indexPath as NSIndexPath).row]
        cell?.textLabel!.text = peripheral.name
        if (peripheral.state == .connected) {
            cell?.backgroundColor = UIColor.init(red: 0, green: 0, blue: 1, alpha: 0.4)
        } else {
            cell?.backgroundColor = UIColor.white
        }
        
        return cell!
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        // Navigation logic may go here. Create and push another view controller.
        tableView.deselectRow(at: indexPath, animated: true)
        
        self.selectedIndex = indexPath
    }
    
    func tableView(_ tableView: UITableView, shouldHighlightRowAt indexPath: IndexPath) -> Bool {
        return true
    }
    
    func tableView(_ tableView: UITableView, didHighlightRowAt indexPath: IndexPath) {
        if  self.selectedIndex != nil && self.selectedIndex.row < self.bluetoothProducts?.count {
            let preCell:UITableViewCell = tableView.cellForRow(at: self.selectedIndex)!
            preCell.backgroundColor = UIColor.white
        }
        
        let cell:UITableViewCell = tableView.cellForRow(at: indexPath)!
        cell.backgroundColor = UIColor.gray
    }
}
